protocol/synthetix/contracts/storage/CollateralConfiguration.sol

```diff
@@ -215,17 +220,33 @@ library CollateralConfiguration {
      * E.g: USDC uses 6 decimals of precision, so this would upscale it by 12 decimals.
      * @param self The CollateralConfiguration object corresponding to the collateral type being converted.
      * @param tokenAmount The token amount, denominated in its native decimal precision.
-     * @return The converted amount, denominated in the system's 18 decimal precision.
+     * @return amountD18 The converted amount, denominated in the system's 18 decimal precision.
      */
     function convertTokenToSystemAmount(
         Data storage self,
         uint256 tokenAmount
-    ) internal view returns (uint256) {
+    ) internal view returns (uint256 amountD18) {
         // this extra condition is to prevent potentially malicious untrusted code from being executed on the next statement
         if (self.tokenAddress == address(0)) {
             revert CollateralNotFound();
         }
 
-        return (tokenAmount * DecimalMath.UNIT) / (10 ** IERC20(self.tokenAddress).decimals());
+        try IERC20(self.tokenAddress).decimals() returns (uint8 decimals) {
+            /// @dev if decimals() is not defined, but an existing fallback function returns a
+            /// bytes representation of uint8, this logic may have unexpected results
+            if (decimals == 0) {
+                amountD18 = tokenAmount;
+            } else if (decimals <= 18) {
+                amountD18 = (tokenAmount * DecimalMath.UNIT) / (10 ** decimals);
+            } else {
+                // ensure no precision is lost when converting to 18 decimals
+                if (tokenAmount % (10 ** DecimalMath.UNIT) != 0) {
```

Comment:
if i'm not mistaken should be `tokenAmount % (10 ** (decimals - DecimalMath.UNIT))`

i'm sure you will pick this up in the tests tho

Reply:
`tokenAmount % (10 ** (decimals - 18))` but yeah, that is correct